File: tutorials/erlang/03-hello-cli/src/hello_cli.erl
-module(hello_cli).
-export([main/1]).

main(Args) ->
    Command = #{
        arguments => [
            #{name => name, type => string, default => "world", long => "-name"},
            #{name => shout, type => boolean, default => false, long => "-shout"}
        ]
    },
    {ok, Parsed, _, _} = argparse:parse(Args, Command),
    Result = hello_cli_format:format(Parsed),
    io:format("~ts~n", [Result]).

File: tutorials/elixir/03-hello-cli/lib/hello_cli.ex
defmodule HelloCli do
  @moduledoc """
  A minimal CLI that greets the user by name.
  """

  def main(argv) do
    argv
    |> parse_args()
    |> format()
    |> IO.puts()
  end

  def parse_args(argv) do
    {opts, _args, _invalid} =
      OptionParser.parse(argv, strict: [name: :string, shout: :boolean])

    %{
      name: Keyword.get(opts, :name, "world"),
      shout: Keyword.get(opts, :shout, false)
    }
  end

  def format(%{name: name, shout: true}) do
    "HELLO, #{String.upcase(name)}!"
  end

  def format(%{name: name}) do
    "Hello, #{name}!"
  end
end
